### PR TITLE
feat: add rule for missing 'provides' method in Deferrable ServiceProviders

### PR DESCRIPTION
**Pull request by PrinsFrank**
- [x] Added or updated tests
- [x] Documented user facing changes
- [x] Updated CHANGELOG.md

**Changes**

This rule will check for a missing 'provides' method in deferrable ServiceProviders.

The 'provides' method is present in the DeferrableProvider interface, but also in the abstract ServiceProvider. If it was not present in the abstract class, this check would not be necessary, but [that PR got rejected](https://github.com/laravel/framework/pull/42460) in Laravel.

**Breaking changes**

### CODE REVIEW

File: src/Rules/DeferrableServiceProviderMissingProvidesRule.php [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+use PhpParser\Node;
+use PHPStan\Analyser\OutOfClassScope;
+use PHPStan\Analyser\Scope;
+use PHPStan\Node\InClassNode;
+use PHPStan\Reflection\MissingMethodFromReflectionException;
+use PHPStan\Rules\Rule;
+use PHPStan\Rules\RuleErrorBuilder;
+use PHPStan\ShouldNotHappenException;
+
+class DeferrableServiceProviderMissingProvidesRule implements Rule
```
Comment (szepeviktor) [inline src/Rules/DeferrableServiceProviderMissingProvidesRule.php:18]: Please add generics here: `@phpstan-implements ...`

File: src/Rules/DeferrableServiceProviderMissingProvidesRule.php [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    /**
+     * @throws ShouldNotHappenException
+     * @throws MissingMethodFromReflectionException
+     */
+    public function processNode(Node $node, Scope $scope): array
+    {
+        /** @var InClassNode $node */
+        $classReflection = $node->getClassReflection();
+        if ($classReflection->isSubclassOf(ServiceProvider::class) === false || $classReflection->implementsInterface(DeferrableProvider::class) === false) {
+            return []; // This rule is only applicable to deferrable serviceProviders
```
Comment (szepeviktor) [inline src/Rules/DeferrableServiceProviderMissingProvidesRule.php:37]:
```suggestion
        // Only applicable to deferrable serviceProviders
        if ($classReflection->isSubclassOf(ServiceProvider::class) === false || $classReflection->implementsInterface(DeferrableProvider::class) === false) {
            return [];
```

File: src/Rules/DeferrableServiceProviderMissingProvidesRule.php [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+        }
+
+        if ($classReflection->hasNativeMethod('provides') === false) {
+            throw new ShouldNotHappenException('If this scenario happens, the "provides" method is removed from the base laravel serviceProvider and this rule can be removed.');
+        }
+
+        $method = $classReflection->getMethod('provides', new OutOfClassScope());
+        if ($method->getDeclaringClass()->getName() !== ServiceProvider::class) {
+            return []; // The provides method is overwritten somewhere in the class hierarchy
```
Comment (szepeviktor) [inline src/Rules/DeferrableServiceProviderMissingProvidesRule.php:46]:
```suggestion
        // The `provides` method is overwritten somewhere in the class hierarchy
        if ($method->getDeclaringClass()->getName() !== ServiceProvider::class) {
            return [];
```

File: src/Rules/DeferrableServiceProviderMissingProvidesRule.php [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+        return InClassNode::class;
+    }
+
+    /**
+     * @throws ShouldNotHappenException
+     * @throws MissingMethodFromReflectionException
+     */
+    public function processNode(Node $node, Scope $scope): array
+    {
+        /** @var InClassNode $node */
```
Comment (szepeviktor) [inline src/Rules/DeferrableServiceProviderMissingProvidesRule.php:34]:
```suggestion
        /** @var \PHPStan\Node\InClassNode $node */
```

File: src/Rules/DeferrableServiceProviderMissingProvidesRule.php [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+class DeferrableServiceProviderMissingProvidesRule implements Rule
+{
+    public function getNodeType(): string
+    {
+        return InClassNode::class;
+    }
+
+    /**
+     * @throws ShouldNotHappenException
+     * @throws MissingMethodFromReflectionException
```
Comment (szepeviktor) [inline src/Rules/DeferrableServiceProviderMissingProvidesRule.php:30]:
```suggestion
     * @throws \PHPStan\ShouldNotHappenException
     * @throws \PHPStan\Reflection\MissingMethodFromReflectionException
```

**Comment by szepeviktor** [comment]
The problem is in https://github.com/nunomaduro/larastan/blob/master/tests/Rules/Data/relation-existence-rule.php

**Comment by szepeviktor** [comment]
The rule name is also in `extension.neon`

**Comment by PrinsFrank** [comment]
@szepeviktor I renamed the test case to confirm this is a test execution issue. That seems to indeed be the case..

File: tests/Rules/AbstractRuleTestCase.php [hunk 6]
```diff
@@ -0,0 +1,17 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Rules;
```
[the remark this answers is not in the document]
Comment (PrinsFrank) [reply tests/Rules/AbstractRuleTestCase.php]: Tx! PHPStorm doesn't understand the namespaces in this project anymore, and neither do I really. Fixed!

File: tests/Rules/CheckDispatchArgumentTypesCompatibleWithClassConstructorRuleTest.php [hunk 7]
```diff
@@ -9,10 +9,10 @@
 use NunoMaduro\Larastan\Rules\CheckDispatchArgumentTypesCompatibleWithClassConstructorRule;
 use PHPStan\Rules\FunctionCallParametersCheck;
 use PHPStan\Rules\Rule;
-use PHPStan\Testing\RuleTestCase;
+use Rules\AbstractRuleTestCase;
```
Comment (szepeviktor) [inline tests/Rules/CheckDispatchArgumentTypesCompatibleWithClassConstructorRuleTest.php:12]:
```suggestion
use Tests\Rules\AbstractRuleTestCase;
```

**Comment by szepeviktor** [comment]
It seems to be green :)
🍏

**Comment by szepeviktor** [comment]
What change in `phpunit.xml.dist` made tests pass?

File: src/Rules/DeferrableServiceProviderMissingProvidesRule.php [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+    public function processNode(Node $node, Scope $scope): array
+    {
+        /** @var \PHPStan\Node\InClassNode $node */
+        $classReflection = $node->getClassReflection();
+        // This rule is only applicable to deferrable serviceProviders
+        if ($classReflection->isSubclassOf(ServiceProvider::class) === false || $classReflection->implementsInterface(DeferrableProvider::class) === false) {
+            return [];
+        }
+
+        if ($classReflection->hasNativeMethod('provides') === false) {
```
Comment (szepeviktor) [inline src/Rules/DeferrableServiceProviderMissingProvidesRule.php:40]:
```suggestion
        if (!$classReflection->isSubclassOf(ServiceProvider::class) || !$classReflection->implementsInterface(DeferrableProvider::class)) {
            return [];
        }

        if (!$classReflection->hasNativeMethod('provides')) {
```

**Comment by PrinsFrank** [comment]
> What change in `phpunit.xml.dist` made tests pass?

None, I added the beStrictAboutChangesToGlobalState setting to try to debug what was going on, but the issue ended up being in the way the container is booted and subsequently cached in the PHPStan RuleTestCase. All extensions need to be loaded or the next time the container is used it is missing dependencies. That's why I added the AbstractRuleTestCase.

**Review by szepeviktor — APPROVED** [review]

**Comment by PrinsFrank** [comment]
@szepeviktor @canvural @nunomaduro I'm cleaning up my forks, can I assume this is not going to get merged and close this PR?

**Comment by PrinsFrank** [comment]
@szepeviktor Ok, np!

**Comment by canvural** [comment]
Thank you! I made some changes to simplify it little bit.

**Pull request merged**